lib/lambda_ethereum_consensus/p2p/block_downloader.ex

```diff
@@ -82,46 +90,71 @@ defmodule LambdaEthereumConsensus.P2P.BlockDownloader do
     with {:ok, stream} <- Libp2p.host_new_stream(host, peer_id, @blocks_by_root_protocol_id),
          :ok <- Libp2p.stream_write(stream, size_header <> compressed_payload),
          :ok <- Libp2p.stream_close_write(stream),
-         {:ok, chunk} <- read_response(stream),
-         {:ok, block} <- decode_response(chunk) do
-      {:ok, block}
+         {:ok, chunks} <- read_response(stream),
+         {:ok, blocks} <- decode_chunks(chunks) do
+      {:ok, blocks}
     else
       {:error, reason} ->
         if retries > 0 do
-          Logger.debug("Retrying request for block with root #{Base.encode16(root)}")
-          request_block_by_root(root, host, retries - 1)
+          Logger.debug(
+            "Retrying request for block with roots #{Enum.map_join(roots, ", ", &Base.encode16/1)}"
+          )
+
+          request_blocks_by_root(roots, host, retries - 1)
         else
           {:error, reason}
         end
     end
   end
 
-  defp read_response(stream) do
-    result =
-      stream
-      |> Libp2p.Stream.from()
-      |> Enum.reduce({:ok, ""}, fn
-        {:ok, chunk}, {:ok, acc} -> {:ok, acc <> chunk}
-        {:error, reason}, _ -> {:error, reason}
-      end)
+  @doc """
+  Reads the response from a stream and returns a list of chunks.
 
+  The stream should consist of the following format
+    response  ::= <response_chunk>*
+    response_chunk  ::= <result> | <context-bytes> | <chunk>
+    result    ::= “0” | “1” | “2” | [“128” ... ”255”]
+
+  We need to ensure that the context bytes correspond to the current fork digest
+  """
+  @spec read_response(Libp2p.stream()) :: {:ok, [binary()]} | {:error, binary()}
+  def read_response(stream) do
     fork_context = @fork_context
 
-    case result do
-      {:ok, ""} ->
-        {:error, "unexpected EOF"}
+    stream
+    |> Libp2p.Stream.from()
+    # Takes a list of N arbitrary binary chunks and returns a list of M <chunk>s
+    # For example:
+    # [<<0, fork_context, 1>>, <<2, 3>>, <<0, fork_context>>, <<4>>, <<5>>] -> [<<1, 2, 3>>, <<4, 5>>]
```

Comment:
Wouldn't this fail when headers aren't at chunk start? For example:

```
[<<0, fork_context, 1>>, <<2, 3, 0>>, <<fork_context>>, <<4, 5>>] -> [<<1, 2, 3, 0, fork_context, 4, 5>>]
```

Reply:
Yes, they would. But Im getting that the chunks from the stream don't come in total arbitrary order, because if that's the case would the Snappy streaming decompression work?